projects/client/RabbitMQ.Client/src/client/impl/Frame.cs

```diff
@@ -215,4 +215,166 @@ public void WriteTo(NetworkBinaryWriter writer)
             writer.Write((byte)Constants.FrameEnd);
         }
     }
+
+    public class AsyncFrame
+    {
+        public MemoryStream m_accumulator;
+
+        public AsyncFrame()
+        {
+        }
+
+        public AsyncFrame(int type, int channel)
+        {
+            Type = type;
+            Channel = channel;
+            Payload = null;
+            m_accumulator = new MemoryStream();
+        }
+
+        public AsyncFrame(int type, int channel, byte[] payload)
+        {
+            Type = type;
+            Channel = channel;
+            Payload = payload;
+            m_accumulator = null;
+        }
+
+        public int Channel { get; set; }
+
+        public byte[] Payload { get; set; }
+
+        public int Type { get; set; }
+
+        public static void ProcessProtocolHeader(NetworkBinaryReader reader)
+        {
+            try
+            {
+                byte b1 = reader.ReadByte();
+                byte b2 = reader.ReadByte();
+                byte b3 = reader.ReadByte();
+                if (b1 != 'M' || b2 != 'Q' || b3 != 'P')
+                {
+                    throw new MalformedFrameException("Invalid AMQP protocol header from server");
+                }
+
+                int transportHigh = reader.ReadByte();
+                int transportLow = reader.ReadByte();
+                int serverMajor = reader.ReadByte();
+                int serverMinor = reader.ReadByte();
+                throw new PacketNotRecognizedException(transportHigh,
+                    transportLow,
+                    serverMajor,
+                    serverMinor);
+            }
+            catch (EndOfStreamException)
+            {
+                // Ideally we'd wrap the EndOfStreamException in the
+                // MalformedFrameException, but unfortunately the
+                // design of MalformedFrameException's superclass,
+                // ProtocolViolationException, doesn't permit
+                // this. Fortunately, the call stack in the
+                // EndOfStreamException is largely irrelevant at this
+                // point, so can safely be ignored.
+                throw new MalformedFrameException("Invalid AMQP protocol header from server");
+            }
+        }
+
+        public static Frame ReadFrom(NetworkBinaryReader reader)
+        {
+            int type;
+
+            try
+            {
+                type = reader.ReadByte();
```

Comment:
Yes. The intention of this PR was never to do something complete. The intent was to explore the impact and show potential road blockers. This to gather ideas how to proceed. It would be a suicide for a "contributor" like me to just send in a large PR without involvement of the maintainers

> Am 02.03.2016 um 21:37 schrieb Joshua Benjamin <notifications@github.com>:
> 
> In projects/client/RabbitMQ.Client/src/client/impl/Frame.cs:
> 
> > +                // design of MalformedFrameException's superclass,
> > +                // ProtocolViolationException, doesn't permit
> > +                // this. Fortunately, the call stack in the
> > +                // EndOfStreamException is largely irrelevant at this
> > +                // point, so can safely be ignored.
> > +                throw new MalformedFrameException("Invalid AMQP protocol header from server");
> > +            }
> > +        }
> > +
> > +        public static Frame ReadFrom(NetworkBinaryReader reader)
> > +        {
> > +            int type;
> > +
> > +            try
> > +            {
> > +                type = reader.ReadByte();
> This is where async / await would play a big role.
> 
> —
> Reply to this email directly or view it on GitHub.
>